hal/src/photon/concurrent_hal.cpp

```diff
@@ -114,11 +103,239 @@ os_result_t os_thread_join(os_thread_t* thread)
  * @param thread    The thread to clean up.
  * @return 0 on success.
  */
-os_result_t os_thread_cleanup(os_thread_t* thread)
+os_result_t os_thread_cleanup(os_thread_t thread)
 {
-    return wiced_rtos_delete_thread((wiced_thread_t*)thread);
+    vTaskDelete(thread);
+    return 0;
 }
 
+/**
+ * Map gthread handles to FreeRTOS task handles.
+ */
+typedef TaskHandle_t __gthread_t;
+
+
+class ThreadQueue
+{
+    QueueHandle_t queue;
+
+public:
+
+    ThreadQueue(UBaseType_t max_size) {
+        queue = xQueueCreate(max_size, sizeof(TaskHandle_t));
+    }
+
+    ~ThreadQueue() {
+        vQueueDelete(queue);
+    }
+
+
+    void enqueue(TicksType_t ticks=portMAX_DELAY) {
+        TaskHandle_t current = xTaskGetCurrentTaskHandle();
+        xQueueSend(queue, &current, ticks);
+    }
+
+    void sleep() {
+        enqueue();
+        vTaskSuspend(NULL);
+    }
+
+    bool wake() {
+        TaskHandle_t next;
+        if (xQueueReceive(queue, &next, 0)==pdTRUE) {
+            vTaskResume(next);
+            return true;
+        }
+        return false;
+    }
+
+    void wakeAll()
+    {
+        while (wake()) {}
+    }
+};
+
+
+class Lock
+{
+    SemaphoreHandle_t semaphore;
+
+public:
+
+    Lock() {
+        semaphore = xSemaphoreCreateMutex();
+    }
+
+    void acquire() {
+        while (xSemaphoreTake(semaphore, portMAX_DELAY)==pdFALSE);
+    }
+
+    void release() {
+        xSemaphoreGive(semaphore);
+    }
+
+    ~Lock() {
+        vSemaphoreDelete(semaphore);
+    }
+};
+
+
+class ConditionVariable
+{
+    ThreadQueue queue;
+
+    typedef std::unique_lock<std::mutex> lock_t;
+public:
+    ConditionVariable(UBaseType_t max_size) : queue(max_size) {}
+
+    void wait(lock_t* lock)
+    {
+        taskENTER_CRITICAL();
+        lock->unlock();
+        queue.enqueue();
+        taskEXIT_CRITICAL();
+        vTaskSuspend(NULL);
+        lock->lock();
+    }
```

Comment:
Glad to help. It take a lot of eyeballs to spot issues with concurrent programming.

One other thought I had: would it be possible to replace the promise in invoke_work by a queue with one item? Create a queue in the caller thread, pass it to the worker thread, when the work completes, the worker puts the result onto the queue. The caller takes from the queue and (potentially) blocks until the result is ready, then deletes the queue (could be handled by shared_ptr). It's less general, but it would allow you to remove promises and condition variables completely from the code (that's a pretty large bit of code).

Right now the only case where you need that promise is when the app blocks on a system call. A promise isn't returned from SYSTEM_THREAD_CONTEXT_ASYNC so you really don't have to worry about the case where the app does a system call and never gets the result of the promise (although it looks like it would work fine because you're using shared_ptr for that reason).

Just a suggestion to try to simplify the task completion synchronization code.